### ISSUE TITLE
Render links and images that work both in md previews as well as the site

### ISSUE THREAD

**Issue by YashJipkate** [11ty/eleventy#1204]
HUGO has this feature using render-hooks wherein the links in the site point to the URLs of the site but in the preview of the MD file from which it is generated has the same link pointing to the relative location of the destination.

I also want to achieve this kind of functionality in 11ty where the links and images are rendered on both platforms since it gives a lot of ease to the editors who can directly edit it on github or preview the links and images without having to set up local environment. But I am not able to find a way to do this anywhere. Any help is appreciated!

Thanks.

**Comment by Ryuno-Ki** [comment]
The link you mentioned in `/docs/copy.md` is a plain old Markdown link:
`[templateFormats configuration](/docs/config/#template-formats)` (section Passthrough File Copy).

As such, it is an absolute link.
If you try to follow it from within GitHub (not GitHub Pages!), it will yield a 404.

**Comment by YashJipkate** [comment]
Hi @Ryuno-Ki, thanks for the correction. I assumed that the links must be working.
I have updated the issue description
Thanks!

**Comment by ejsmith** [comment]
I'm running into this as well. When using a tool like vscode to edit my markdown files I can get things like intellisense when my links and images go to physically relative paths, but then those links do not work in the rendered site because myfile.md gets moved into /myfile/ on the rendered site. Also, if I take the content from that markdown file and show it in a listing page like a main blog feed then the same content would be rendered to something like /news/ and the relative links are broken there as well. It would be really nice if eleventy had the ability to take relative paths in content (content links, images, etc) and turn them into fully qualified paths in the rendered site content.

**Comment by ejsmith** [comment]
This would also be good to enable linting tools to verify links.

**Comment by nhoizey** [comment]
Here's how I do it:

# My source folder hierarchy is the same as he built one

For example:
- Source: https://github.com/nhoizey/nicolas-hoizey.com/tree/master/src/articles/2020/10/26/enhancing-archives-navigation-step-1/index.md
- Build: https://nicolas-hoizey.com/articles/2020/10/26/enhancing-archives-navigation-step-1/index.html

# I copy all images from the source to the build folder, keeping the same paths

```javascript
  eleventyConfig
    .addPassthroughCopy(
      path.join(config.dir.src, '**/*.{jpg,png,gif,svg,kmz,zip,css}')
    );
```

So for this image:
- Source: https://github.com/nhoizey/nicolas-hoizey.com/blob/master/src/articles/2020/10/26/enhancing-archives-navigation-step-1/months-pagination-after.png
- Build: https://nicolas-hoizey.com/articles/2020/10/26/enhancing-archives-navigation-step-1/months-pagination-after.png

# I use relative URLs for images in Markdown

As you might have already seen in [the Markdown source of the page](https://raw.githubusercontent.com/nhoizey/nicolas-hoizey.com/master/src/articles/2020/10/26/enhancing-archives-navigation-step-1/index.md), here's how I add the image:

```markdown
![A much more user friendly months navigation](months-pagination-after.png)
```

It means:
- any local editor (I use both VS Code and Macdown) with preview can show the image
- the page in Github can show the image:  https://github.com/nhoizey/nicolas-hoizey.com/tree/master/src/articles/2020/10/26/enhancing-archives-navigation-step-1/index.md

**Comment by montanaflynn** [comment]
How about for actual markdown links? 

```markdown
# README

Linking to [DOCS](./DOCS.md) page.
```

Changes to 

```html
<h1>README</h1>
<p>Linking to <a href="./DOCS.md">DOCS</a> page.</p>
```

This means the link doesn't work. Surely there is a way to have both markdown and html links stay working?

Related issue: https://github.com/11ty/eleventy/issues/976

**Comment by nhoizey** [comment]
@montanaflynn I guess you can use links to folders only, and use `README.md` files that Github will render. And use `index.html` as Eleventy permalink for `README.md` files.

**Comment by ejsmith** [comment]
I added these `markdownItOptions` and it seems to work pretty well:

```js
  const markdownItOptions = {
    html: true,
    breaks: true,
    linkify: false,
    replaceLink: function (link, env) {
      if (link.startsWith('http://')
        || link.startsWith('https://')
        || link.startsWith('/')
        || link.startsWith('#')
        || link.startsWith('mailto:'))
        return link;
      
      let dir = path.dirname(env.page?.filePathStem ?? '');
      let fullPath = dir + '/' + link;
      
      fullPath = fullPath.replace(/index.md$/, '').replace(/.md$/, '/');
      
      return fullPath;
    }
  };
```

**Comment by Ryuno-Ki** [comment]
```js
      if (link.startsWith('http://')
        || link.startsWith('https://')
        || link.startsWith('/')
        || link.startsWith('#')
        || link.startsWith('mailto:')) {
        return link;
}
```
could be turned into
```js
const doNothing = ['http://', 'https://', '/', '#', 'mailto:'];
if (doNothing.some((protocol) => link.startsWith(protocol))) { return link; }
```
:-)

**Comment by montanaflynn** [comment]
@nhoizey it's not so much about GitHub but my editor VSCode which can preview and follow markdown links. Also other tools like Obsidian and other markdown editors have this functionality.

@ejsmith @Ryuno-Ki thanks I played around with those but it wasn't working as well as I would like for various link styles. I ended up modifying it to this which works for all the links I have tested.

```js
  let markdownItOptions = {
    html: true,
    breaks: true,
    linkify: true,
    replaceLink: function (link, env) {
      const doNothing = ['http://', 'https://', '#', 'mailto:'];
      if (doNothing.some((protocol) => link.startsWith(protocol))) { return link; }

      // Convert links such as `/page.md` to `/page/`
      if (link.startsWith("/")) {
        const extensions = ['md', 'mmd', 'mdl', 'markdown']
        if (extensions.some((ext) => link.endsWith(ext))) { 
          const regex = /(?!(?:http:|https:))(.*?)(.md)/gm;
          const newLink = link.replace(regex, "$1")
          console.log(link, newLink)
          return newLink
        }
      }

      // Convert links such as `./page.md` to `.././page/`
      const extensions = ['md', 'mmd', 'mdl', 'markdown']
      if (extensions.some((ext) => link.endsWith(ext))) { 
        const regex = /(?!(?:http:|https:))(.*?)(.md)/gm;
        const newLink = link.replace(regex, "../$1")
        console.log(link, newLink)
        return newLink
      }
      return link
    }
  };
  
  let md = markdownIt(markdownItOptions)
```

Also note you need to install the `markdown-it-replace-link` package and use it as well:

```js
const markdownItReplaceLink = require('markdown-it-replace-link');
md.use(markdownItReplaceLink)
````

**Comment by nhoizey** [comment]
@montanaflynn ok, I thought it was also about GitHub.

**Comment by victornpb** [comment]
I just made this plugin to solve this limitation
https://www.npmjs.com/package/eleventy-plugin-page-assets

**Comment by Apostolique** [comment]
I'm testing this version:

```js
let config = {
  dir: {
    input: 'docs',
  },
}
```

```js
    replaceLink(link, env) {
      // Skip outbound link.
      let isOutbound = new RegExp('^(?:[a-z]+:)?//', 'i');
      if (isOutbound.test(link)) { return link }

      // Remove markdown extension.
      let isMarkdown = new RegExp('(?:readme)?.(md|mmd|mdl|markdown)$', 'i');
      if (isMarkdown.test(link)) {
        link = link.replace(isMarkdown, "");
      }

      // Check if there's an extension.
      let hasExt = /\.([0-9a-z]+)(?:[\?#]|$)/i;
      if (!hasExt.test(link)) {
        // Add trailing slash if missing.
        link = link.replace(/\/?(\?|#|$)/, '/$1');
      }

      // Fix absolute urls. (Might need a check if the link points somewhere in the repo outside the docs.)
      let isAbsolute = new RegExp(`^\/${config.dir.input}`);
      if (isAbsolute.test(link)) {
        link = link.replace(isAbsolute, "");
        return eleventyConfig.getFilter("url")(link);
      }

      // Fix relative links.
      return eleventyConfig.getFilter("url")(`../${link}`);
    }
```

I set up my eleventy to handle `README.md` files as if they were `index.md` files.

Seems to handle a lot of edge cases so far though I'm still testing to find bugs.

**Comment by montanaflynn** [comment]
@victornpb does it rewriting links to other markdown files as well? Looks like it's mostly dealing with assets, I usually just turn on `setPassthroughAll` so all assets are moved to the same folders they were in and then referencing assets works for me. Wondering what the plugin's use case is?

@Apostolique I'll test this out as well, I noticed a lot of edge cases in my comment's code especially with nested pages linking to parents.

```
content
├── DOCS.md
├── INDEX.md
├── README.md
├── SYNTAX.mmd
├── screenshot.png
└── CLI
    ├── DOCS.md
    └── README.md
```

May I ask how you set up eleventy to handle README.md as if they were index.md files? I also had that set up along with lowercasing all output files as a collection filter but it has some issues as well.

```js
  eleventyConfig.addCollection('pages', collection => {
    return collection.getAllSorted().map(item => {
      item.outputPath = item.outputPath.replace('README.html', 'INDEX.html');
      item.outputPath = item.outputPath.toLowerCase();
    });
  });
```

Ideally there would be some eleventy options so if there is a readme but not index to treat it like the index. And also to lowercase all the output since I don't think anyone really wants mixed case or uppercase URLs.

I also want eleventy to work with both lowercase and uppercase input files and had to modify eleventy's `./src/TemplateFileSlug.js` file to work with both `INDEX.md` and `index.md` files.

**Comment by Apostolique** [comment]
@montanaflynn I used this solution: https://github.com/11ty/eleventy/issues/1379#issuecomment-680952805

```js
  eleventyConfig.addCollection('readme', collection => {
    return collection.getAllSorted().map(item => {
      if (item.fileSlug === 'README') {
        item.outputPath = item.outputPath.replace('/README', '');
      }
    });
  });
```

I didn't bother handling the lowercase but I guess I could since I already hijacked the link generation.

**Comment by montanaflynn** [comment]
@Apostolique thanks, I did find a problem with your replaceLink function:

Linking to a nested index file doesn't work and likewise linking from an index to a sibling also doesn't work.

```
content
├── index.md
├── sibling.md
└── cli
    ├── index.md
    └── sibling.md
```

```
# ./cli/index.md

Here's a [sibling](./sibling.md)
```

```
# ./cli/sibling.md

Here's cli [index](./index.md)
```

**Comment by Apostolique** [comment]
@montanaflynn I fixed the link from sibling.md to index.md. I'm not yet sure how to fix from index.md (or README.md) to sibling.md.

It fix it, I need to change the relative line:
```js
      // Fix relative links.
      return eleventyConfig.getFilter("url")(`../${link}`);
```

To something like:
```js
      if (isInsideReadmeOrIndex) {
         return eleventyConfig.getFilter("url")(link);
      } else {
         // Fix relative links.
         return eleventyConfig.getFilter("url")(`../${link}`);
      }
```

To do that though, I need to know what's the url or slug of the current file from inside the markdown parser. Most likely I would need access to the page variables: <https://www.11ty.dev/docs/data-eleventy-supplied/#page-variable-contents>.

**Comment by montanaflynn** [comment]
@Apostolique you're the best! One more edge case: linking to `/` from a nested folder goes to the nested folder not the root.

```
# ./cli/sibling.md
Linking to [home page](/).
```

**Comment by Apostolique** [comment]
I wasn't able to reproduce that one. It links to the home page for me. It's handled here:
```js
      // Fix absolute url. (Might need a check if the link points somewhere in the repo outside the docs.)
      let isAbsolute = new RegExp(`^\/${config.dir.input}`);
      if (isAbsolute.test(link)) {
        link = link.replace(isAbsolute, "");
        return eleventyConfig.getFilter("url")(link);
      }
```
It follows the url filter from: <https://www.11ty.dev/docs/filters/url/>.

**Comment by montanaflynn** [comment]
@Apostolique it's not getting caught for me because it doesn't match the `^\/${config.dir.input}` regex.

if `config.dir.input = 'docs'` it would only match `/docs` but not simply `/` right?

Also I don't think having to hard code the config input is suitable since it's possible to use the command line options for input rather than putting it in `.eleventy.js`. I couldn't find how to get that dynamically from `eleventyConfig` or `env` either.

**Comment by Apostolique** [comment]
The problem is that if your input dir is `docs`, then your home page's link should be `/docs/` in the markdown.

The other links outside the docs folder should become links to the public repo. Here is a [cheat sheet](https://github.com/sm-artlight/github-cheatsheet) for GitHub markdown links.

**Comment by zachleat** [comment]
Cleanup up some old issues—thanks everyone!

**Comment by warpfork** [comment]
Did this get any built-in upstream support in recent 11ty releases?  If so, I would super love a pointer.

I just saw someone merge things into an 11ty site I launched which adds code "based on" snippets seen in this issue, or linked from here, but that's... Well, I'm glad it got the feature improvement, but this is such a common issue that I'd rather not cargo-cult that code into new 11ty projects if there's an alternative.  (Especially because it includes an awful lot of regexps...)
